File: client/src/types/render.ts
export type VisualTemplate =
  | "nebula"
  | "geometric"
  | "waveform"
  | "cinematic"
  | "retro"
  | "nature"
  | "abstract"
  | "urban"
  | "glitchbreak"
  | "90s-anime";

export type MotionStyle =
  | "slow-drift"
  | "pulse"
  | "energetic"
  | "chaotic"
  | "breathing"
  | "glitch"
  | "smooth-flow"
  | "staccato";

export type TransitionType =
  | "fade-from-black"
  | "fade-to-black"
  | "cross-dissolve"
  | "hard-cut"
  | "morph"
  | "flash-white"
  | "wipe"
  | "zoom-in"
  | "zoom-out";

export type LyricsAnimation = "fade-word" | "typewriter" | "karaoke" | "float-up" | "none";

export interface LyricsDisplayConfig {
  enabled: boolean;
  font: "sans" | "serif" | "mono";
  size: "small" | "medium" | "large";
  animation: LyricsAnimation;
  color: string;
  shadow: boolean;
}

export interface SectionSpec {
  label: string;
  startTime: number;
  endTime: number;
  colorPalette: string[];
  motionStyle: MotionStyle;
  intensity: number;
  aiPrompt: string;
  transitionIn: TransitionType;
  transitionOut: TransitionType;
  visualElements: string[];
  keyframeUrl?: string;
}

export interface GlobalStyle {
  template: VisualTemplate;
  styleModifiers: string[];
  recurringMotifs: string[];
  lyricsDisplay: LyricsDisplayConfig;
}

export type AspectRatio = "16:9" | "9:16" | "1:1";
export type VideoQuality = "draft" | "standard" | "high";

export interface ExportSettings {
  resolution: [number, number];
  fps: 24 | 30 | 60;
  aspectRatio: AspectRatio;
  format: "mp4";
  quality: VideoQuality;
}

export interface RenderSpec {
  globalStyle: GlobalStyle;
  sections: SectionSpec[];
  exportSettings: ExportSettings;
}

export type ExportPreset =
  | "youtube"
  | "youtube-hd"
  | "tiktok"
  | "instagram"
  | "twitter"
  | "4k"
  | "custom";

export interface ExportPresetConfig {
  label: string;
  resolution: [number, number];
  aspectRatio: AspectRatio;
  fps: 24 | 30 | 60;
  description: string;
}

export const EXPORT_PRESETS: Record<ExportPreset, ExportPresetConfig> = {
  youtube: {
    label: "YouTube",
    resolution: [1920, 1080],
    aspectRatio: "16:9",
    fps: 30,
    description: "YouTube, Vimeo",
  },
  "youtube-hd": {
    label: "YouTube HD",
    resolution: [1920, 1080],
    aspectRatio: "16:9",
    fps: 60,
    description: "YouTube high frame rate",
  },
  tiktok: {
    label: "TikTok / Reels",
    resolution: [1080, 1920],
    aspectRatio: "9:16",
    fps: 30,
    description: "TikTok, Instagram Reels, Shorts",
  },
  instagram: {
    label: "Instagram Square",
    resolution: [1080, 1080],
    aspectRatio: "1:1",
    fps: 30,
    description: "Instagram Feed",
  },
  twitter: {
    label: "Twitter / X",
    resolution: [1280, 720],
    aspectRatio: "16:9",
    fps: 30,
    description: "Twitter/X video",
  },
  "4k": {
    label: "4K",
    resolution: [3840, 2160],
    aspectRatio: "16:9",
    fps: 30,
    description: "High quality archive",
  },
  custom: {
    label: "Custom",
    resolution: [1920, 1080],
    aspectRatio: "16:9",
    fps: 30,
    description: "User-defined settings",
  },
};

File: server/app/models/render.py
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


def _to_camel(string: str) -> str:
    """Convert snake_case to camelCase."""
    components = string.split("_")
    return components[0] + "".join(x.title() for x in components[1:])


VisualTemplate = Literal[
    "shader", "nebula", "geometric", "waveform", "cinematic", "retro", "nature", "abstract", "urban",
    "glitchbreak", "90s-anime",
]
MotionStyle = Literal[
    "slow-drift", "pulse", "energetic", "chaotic", "breathing", "glitch", "smooth-flow", "staccato"
]
TransitionType = Literal[
    "fade-from-black",
    "fade-to-black",
    "cross-dissolve",
    "hard-cut",
    "morph",
    "flash-white",
    "wipe",
    "zoom-in",
    "zoom-out",
]
LyricsAnimation = Literal["fade-word", "typewriter", "karaoke", "float-up", "none"]
AspectRatio = Literal["16:9", "9:16", "1:1"]
VideoQuality = Literal["draft", "standard", "high"]


class LyricsDisplayConfig(BaseModel):
    model_config = ConfigDict(populate_by_name=True, alias_generator=_to_camel)

    enabled: bool = True
    font: Literal["sans", "serif", "mono"] = "sans"
    size: Literal["small", "medium", "large"] = "medium"
    animation: LyricsAnimation = "fade-word"
    color: str = "#F0F0F5"
    shadow: bool = True


class SectionSpec(BaseModel):
    model_config = ConfigDict(populate_by_name=True, alias_generator=_to_camel)

    label: str
    start_time: float
    end_time: float
    color_palette: list[str] = Field(default_factory=lambda: ["#7C5CFC", "#1A1A28"])
    motion_style: MotionStyle = "slow-drift"
    intensity: float = Field(ge=0, le=1, default=0.5)
    ai_prompt: str = ""
    transition_in: TransitionType = "cross-dissolve"
    transition_out: TransitionType = "cross-dissolve"
    visual_elements: list[str] = Field(default_factory=list)
    keyframe_url: str | None = None


class GlobalStyle(BaseModel):
    model_config = ConfigDict(populate_by_name=True, alias_generator=_to_camel)

    template: VisualTemplate = "shader"
    shader_description: str = ""
    style_modifiers: list[str] = Field(default_factory=list)
    recurring_motifs: list[str] = Field(default_factory=list)
    lyrics_display: LyricsDisplayConfig = Field(default_factory=LyricsDisplayConfig)


class ExportSettings(BaseModel):
    model_config = ConfigDict(populate_by_name=True, alias_generator=_to_camel)

    resolution: tuple[int, int] = (1920, 1080)
    fps: Literal[24, 30, 60] = 30
    aspect_ratio: AspectRatio = "16:9"
    format: Literal["mp4"] = "mp4"
    quality: VideoQuality = "high"


class RenderSpec(BaseModel):
    model_config = ConfigDict(populate_by_name=True, alias_generator=_to_camel)

    global_style: GlobalStyle = Field(default_factory=GlobalStyle)
    sections: list[SectionSpec] = Field(default_factory=list)
    export_settings: ExportSettings = Field(default_factory=ExportSettings)


class RenderRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True, alias_generator=_to_camel)

    job_id: str
    render_spec: RenderSpec


class RenderEditRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True, alias_generator=_to_camel)

    edit_description: str
    render_spec: RenderSpec | None = None
